toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -229,6 +232,31 @@ class DiagnosticEmitter {
       annotate_fns_;
 };
 
+// This relies on `void*` location handling on `DiagnosticEmitter`.
+class NoLocDiagnosticEmitter : public DiagnosticEmitter<void*> {
```

Comment:
Possible TODO to add here:

```suggestion
// This relies on `void*` location handling on `DiagnosticEmitter`.
//
// TODO: Based on how this ends up used or if we get more distinct
// emitters, it might be worth considering having diagnostics
// specify that they don't apply to source-location carrying
// emitters. For example, this might look like a
// `CARBON_NO_LOC_DIAGNOSTIC` macro, or some other factoring.
// But it might end up being more noise than it is worth.
class NoLocDiagnosticEmitter : public DiagnosticEmitter<void*> {
```